ha/core/event_manager/event_manager.py

```diff
@@ -57,42 +89,181 @@ def _validate_component(self, component: str) -> None:
         if component.lower() not in SUBSCRIPTION_LIST:
             raise InvalidComponent(f"Invalid component {component}, not part of subscription list.")
 
-    def _validate_events(self, events: list) -> None:
-        """
-        Raise error InvalidEvent if event is not valid.
+    def subscribe(self, component : str, events : list):
+        '''
+           Register all the events for the notification. It maintains
+           list of events registered by the components. It maps the event
+           name with component name and store it in the consul. \
+           Also, maps the component with event
 
-        Args:
-            events (list): Event list.
-        """
-        if not isinstance(events, list):
-            raise InvalidEvent(f"Invalid type {events}, event type should be list")
-        for event in events:
-            if event not in EVENTS:
-                raise InvalidEvent(f"Invalid event: {event}, not part of HA event list.")
-        Log.debug(f"event: {event} is valid for subscription request")
-
-    def subscribe(self, component: str, events: list) -> str:
-        """
-        Register events for the notification. It maintains list of events registered by the components.
-
-        Args:
+           Args:
             component (str): Component Name
             events (list): Events.
-        """
-        pass
+        '''
+        Log.info(f'Received a subscribe request from {component} for event: {events}')
+        EventManager._validate_component(component)
+        EventManager._validate_events(events)
+        if component:
+            message_type = self._create_message_type(component)
+        resource_type = events.resource_type
+        states = events.states
+        self._store_component_key(f'{const.COMPONENT_KEY}/{component}', resource_type, states)
+        self._store_event_key(f'{const.EVENT_KEY}', resource_type, states, comp=component)
+        Log.info('Subscription request is successfully stored into consul')
+        return message_type
+
+    def _store_component_key(self, key: str, resource_type: str, val: list) -> None:
+        '''
+           Perform actual consul store operation for component keys
+           Ex:
+           key: /cortx/ha/v2/events/subscribe/sspl
+           confstore value: ['enclosure:sensor:voltage/failed', ...]
+        '''
+        if self._confstore.key_exists(key):
+            comp_json_list = self._confstore.get(key)
 
-    def unsubscribe(self, component: str, events: list = None) -> None:
+            # Get the compnent list from consul first
+            comp_list = json.loads(comp_json_list['cortx/ha/v1/' + f'{key}'])
```

Comment:
I think we can use key_val.popitem() to get value from the key_val returned by confstore.get() above.

Reply:
Done